sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala

```diff
@@ -56,16 +68,27 @@ case class OptimizeLocalShuffleReader(conf: SQLConf) extends Rule[SparkPlan] {
         case e: ShuffleExchangeExec => e
       }.length
     }
-
+    // Check whether additional shuffle introduced. If introduced, revert the local reader.
     val numExchangeBefore = numExchanges(EnsureRequirements(conf).apply(plan))
-    val numExchangeAfter = numExchanges(EnsureRequirements(conf).apply(optimizedPlan))
-
-    if (numExchangeAfter > numExchangeBefore) {
-      logDebug("OptimizeLocalShuffleReader rule is not applied due" +
+    val numExchangeAfter = numExchanges(EnsureRequirements(conf).apply(withProbeSideLocalReader))
+    val optimizedPlan = if (numExchangeAfter > numExchangeBefore) {
+      logDebug("OptimizeLocalShuffleReader rule is not applied in the probe side due" +
         " to additional shuffles will be introduced.")
       plan
     } else {
-      optimizedPlan
+      withProbeSideLocalReader
+    }
+    // Add the local reader in build side and and do not need to check whether
+    // additional shuffle introduced.
+    optimizedPlan.transformDown {
+      case join: BroadcastHashJoinExec if (join.buildSide == BuildLeft &&
```

Comment:
nit: we can unify the condition check some how
```
object BroadcastJoinWithShuffleLeft {
  def unapply(plan: LogicalPlan): Option[(QueryStageExec, BuildSide)] = plan match {
    case join: BroadcastHashJoinExec if ShuffleQueryStageExec.isShuffleQueryStageExec(join.left) =>
      Some((join.left.asInstanceOf[QueryStageExec], join.buildSide))
  }
}
object BroadcastJoinWithShuffleRight ...

// add probe side local reader
case join @ BroadcastJoinWithShuffleLeft(shuffleStage, BuildRight) =>
  val localReader = ...
  join.copy(left = localReader)
...

// add build side local reader
case join @ BroadcastJoinWithShuffleLeft(shuffleStage, BuildLeft) =>
  ...
```

Reply:
Shall we check for `BroadcastQueryStageExec` for which we shouldn't add a local shuffle reader?

Another alternative is we apply this part of the rule after `EnsureRequirements` and match `BroadcastExchangeExec` over `QueryStageExec` instead.